Modules/_interpretersmodule.c

```diff
@@ -0,0 +1,1892 @@
+        return -1;
+    }
+    return cid;
+}
+
+/* data-sharing-specific code ***********************************************/
+
+struct _shareditem {
+    Py_UNICODE *name;
+    Py_ssize_t namelen;
+    _PyCrossInterpreterData data;
+};
+
+void
+_sharedns_clear(struct _shareditem *shared)
+{
+    for (struct _shareditem *item=shared; item->name != NULL; item += 1) {
+        _PyCrossInterpreterData_Release(&item->data);
+    }
+}
+
+static struct _shareditem *
+_get_shared_ns(PyObject *shareable, Py_ssize_t *lenp)
+{
+    if (shareable == NULL || shareable == Py_None)
+        return NULL;
+    Py_ssize_t len = PyDict_Size(shareable);
+    *lenp = len;
+    if (len == 0)
+        return NULL;
+
+    struct _shareditem *shared = PyMem_NEW(struct _shareditem, len+1);
+    for (Py_ssize_t i=0; i < len; i++) {
+        *(shared + i) = (struct _shareditem){0};
+    }
+    Py_ssize_t pos = 0;
+    for (Py_ssize_t i=0; i < len; i++) {
+        PyObject *key, *value;
+        if (PyDict_Next(shareable, &pos, &key, &value) == 0) {
+            break;
+        }
+        struct _shareditem *item = shared + i;
+
+        if (_PyObject_GetCrossInterpreterData(value, &item->data) != 0)
+            break;
+        item->name = PyUnicode_AsUnicodeAndSize(key, &item->namelen);
+        if (item->name == NULL) {
+            _PyCrossInterpreterData_Release(&item->data);
+            break;
+        }
+        (item + 1)->name = NULL;  // Mark the next one as the last.
+    }
+    if (PyErr_Occurred()) {
+        _sharedns_clear(shared);
+        PyMem_Free(shared);
+        return NULL;
+    }
+    return shared;
+}
+
+static int
+_shareditem_apply(struct _shareditem *item, PyObject *ns)
+{
+    PyObject *name = PyUnicode_FromUnicode(item->name, item->namelen);
+    if (name == NULL) {
+        return 1;
+    }
+    PyObject *value = _PyCrossInterpreterData_NewObject(&item->data);
+    if (value == NULL) {
+        Py_DECREF(name);
+        return 1;
+    }
+    int res = PyDict_SetItem(ns, name, value);
+    Py_DECREF(name);
+    Py_DECREF(value);
+    return res;
+}
+
+// Ultimately we'd like to preserve enough information about the
+// exception and traceback that we could re-constitute (or at least
+// simulate, a la traceback.TracebackException), and even chain, a copy
+// of the exception in the calling interpreter.
+
+struct _shared_exception {
+    char *msg;
+};
+
+static struct _shared_exception *
+_get_shared_exception(void)
+{
+    struct _shared_exception *err = PyMem_NEW(struct _shared_exception, 1);
+    if (err == NULL)
+        return NULL;
+    PyObject *exc;
+    PyObject *value;
+    PyObject *tb;
+    PyErr_Fetch(&exc, &value, &tb);
+    PyObject *msg;
+    if (value == NULL)
+        msg = PyUnicode_FromFormat("%S", exc);
+    else
+        msg = PyUnicode_FromFormat("%S: %S", exc, value);
+    if (msg == NULL) {
+        err->msg = "unable to format exception";
```

Comment:
Redundant?  And need to free `err`?

Reply:
I'd meant to return `err` on the next line rather than NULL. :)